### PR TITLE
LLVM 4 Merge

### PR DESCRIPTION
**Pull request by jcarlson23**
I tested this with fi_tests under PTABen and while there were 4 failures (I think, it was just handful for sure) I don't think (although I'm not entirely sure) these were unexpected (non-default cases referred to [Questions on PTABen test cases](https://github.com/unsw-corg/PTABen/issues/2) ).  I ran a few other tests and had generally positive results as well so, I didn't run them all (On my Linux VM this went pretty smooth but on my Mac I'm still struggling a bit although I did find some of my issues, perhaps the key ones).  

This has the BridgedGEPIterator solution implemented along with the previous PR's more mundane fixes, again the NodeRef inclusion in the GraphTraits was the most notable as I recall, to update SVF with LLVM 4.

### CODE REVIEW

**Comment by rockysui** [comment]
Hi Jared,

Thanks! I will take a look at it today.

**Comment by rockysui** [comment]
Your PR works very well on small test cases including `fi_tests` and `fs_tests`. I will start testing it using large benchmarks cpu2000/2006 by comparing the statistics between the existing repo and new one after PR (e.g., points-to size). If everything works, we can merge it asap.

File: lib/MemoryModel/LocMemModel.cpp [hunk 1]
```diff
@@ -60,7 +60,7 @@ bool LocSymTableInfo::computeGepOffset(const llvm::User *V, LocationSet& ls) {
         if (index <= baseIndex) {
             /// variant offset
             // Handling pointer types
-            if (const PointerType* pty = dyn_cast<PointerType>(*gi)) {
+	  if (const PointerType* pty = dyn_cast<PointerType>(gi.getIndexedType())) {
```
Comment (rockysui) [inline lib/MemoryModel/LocMemModel.cpp]: For LocMemModel, we can also get rid of the `gep_type_iterator` and `gi.getIndexedType()`, and stick with your BridgedGepIterator for now:).

Please use `*gi` intead of `gi.getIndexedType()` in LocMemModel. Please also replace `gep_type_iterator` at lines 50 and 57 with the BridgedGepIterator.

File: lib/MemoryModel/LocMemModel.cpp [hunk 2]
```diff
@@ -94,19 +94,19 @@ bool LocSymTableInfo::computeGepOffset(const llvm::User *V, LocationSet& ls) {
...
             }
             // Calculate the size of the array element
-            else if(const ArrayType* at = dyn_cast<ArrayType>(*gi)) {
+            else if(const ArrayType* at = dyn_cast<ArrayType>(gi.getIndexedType())) {
                 const Type* et = at->getElementType();
                 Size_t sz = getTypeSizeInBytes(et);
                 ls.offset += idx * sz;
             }
             // Handling struct here
-            else if (const StructType *ST = dyn_cast<StructType>(*gi)) {
```
Comment (rockysui) [inline lib/MemoryModel/LocMemModel.cpp]: Shall we use `gi.getStructTypeOrNull`? I am not very sure whether these two are equivalent or not. Please keep the original one as you did in `MemModel.cpp`

Comment (jcarlson23) [inline lib/MemoryModel/LocMemModel.cpp]: LocMemModel.cpp updated!

File: lib/MemoryModel/LocMemModel.cpp [hunk 3]
```diff
@@ -94,19 +95,19 @@ bool LocSymTableInfo::computeGepOffset(const llvm::User *V, LocationSet& ls) {
             // Handling pointer types
             // These GEP instructions are simply making address computations from the base pointer address
             // e.g. idx1 = (char*) &MyVar + 4,  at this case gep only one offset index (idx)
-            if (const PointerType* pty = dyn_cast<PointerType>(*gi)) {
+            if (const PointerType* pty = dyn_cast<PointerType>(gi.getIndexedType())) {
```
Comment (rockysui) [inline lib/MemoryModel/LocMemModel.cpp]: Please change this back.

File: lib/MemoryModel/LocMemModel.cpp [hunk 4]
```diff
@@ -94,19 +95,19 @@ bool LocSymTableInfo::computeGepOffset(const llvm::User *V, LocationSet& ls) {
...
                 const Type* et = pty->getElementType();
                 Size_t sz = getTypeSizeInBytes(et);
                 ls.offset += idx * sz;
             }
             // Calculate the size of the array element
-            else if(const ArrayType* at = dyn_cast<ArrayType>(*gi)) {
```
Comment (rockysui) [inline lib/MemoryModel/LocMemModel.cpp]: See above.

File: lib/MemoryModel/LocMemModel.cpp [hunk 5]
```diff
@@ -94,19 +95,19 @@ bool LocSymTableInfo::computeGepOffset(const llvm::User *V, LocationSet& ls) {
...
+            else if(const ArrayType* at = dyn_cast<ArrayType>(gi.getIndexedType())) {
```
Comment (rockysui) [inline lib/MemoryModel/LocMemModel.cpp]: Should be this line.

Comment (jcarlson23) [inline lib/MemoryModel/LocMemModel.cpp]: Sorry, my bad on missing this one, trying to update the commits in between breakfast ;-P

**Comment by rockysui** [comment]
Jared,

Please link your account to your commits, so we can acknowledge your contribution:

https://github.com/unsw-corg/SVF/commit/9ef39817188f95ab13a0e1db4466aa1e9d192277

Thanks

**Comment by jcarlson23** [comment]
Thanks @rockysui!  I think I just took care of that (added the additional email).  Thanks again.

**Pull request merged**